tests/runtime/test_http_server.py

```diff
@@ -0,0 +1,241 @@
+            )
+
+            tls_config = TLSConfig(
+                server=KeyPair(cert=server_certfile, key=server_keyfile),
+                client=KeyPair(cert="", key=""),
+            )
+            # need to save this ca_certfile in config_overrides so the tls tests below can access it from client side
+            config_overrides["use_in_test"] = {"ca_cert": ca_certfile}
+            if mtls:
+                client_certfile, client_keyfile = save_key_cert_pair(
+                    "client",
+                    workdir,
+                    *tls_test_tools.generate_derived_key_cert_pair(ca_key=ca_key),
+                )
+                tls_config.client = KeyPair(cert=ca_certfile, key="")
+                # need to save the client cert and key in config_overrides so the mtls test below can access it
+                config_overrides["use_in_test"]["client_cert"] = client_certfile
+                config_overrides["use_in_test"]["client_key"] = client_keyfile
+
+            config_overrides["runtime"] = {"tls": tls_config}
+        port = http_server.RuntimeServerBase._find_port()
+        config_overrides.setdefault("runtime", {})["http"] = {
+            "server_shutdown_grace_period_seconds": 0.01,  # this is so the server is killed after 0.1 if no test is running
+            "port": port,
+            **http_config_overrides,
+        }
+
+        with temp_config(config_overrides, "merge"):
+            yield config_overrides
+
+
+## Insecure and TLS Tests #######################################################################
+
+
+def test_insecure_server():
+    with generate_tls_configs():
+        insecure_http_server = http_server.RuntimeHTTPServer()
+        # start a non-blocking http server
+        with insecure_http_server:
+            resp = requests.get(f"http://localhost:{insecure_http_server.port}/docs")
+            resp.raise_for_status()
+
+
+def test_basic_tls_server():
+    with generate_tls_configs(
+        tls=True, mtls=False, http_config_overrides={}
+    ) as config_overrides:
+        http_server_with_tls = http_server.RuntimeHTTPServer(
+            tls_config_override=config_overrides["runtime"]["tls"]
+        )
+        # start a non-blocking http server with basic tls
+        with http_server_with_tls:
+            resp = requests.get(
+                f"https://localhost:{http_server_with_tls.port}/docs",
+                verify=config_overrides["use_in_test"]["ca_cert"],
+            )
+            resp.raise_for_status()
+
+
+def test_mutual_tls_server():
+    with generate_tls_configs(
+        tls=True, mtls=True, http_config_overrides={}
+    ) as config_overrides:
+        http_server_with_mtls = http_server.RuntimeHTTPServer(
+            tls_config_override=config_overrides["runtime"]["tls"]
+        )
+        # start a non-blocking http server with mutual tls
+        with http_server_with_mtls:
+            resp = requests.get(
```

Comment:
It would be good to verify that if the client does NOT send a valid client cert it gets rejected as well